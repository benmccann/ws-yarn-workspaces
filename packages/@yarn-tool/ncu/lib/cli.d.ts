/**
 * Created by user on 2020/6/12.
 */
import { Argv } from 'yargs';
export declare function setupNcuToYargs<T extends any>(yargs: Argv<T>): Argv<T & {
    dep: string;
} & {
    minimal: boolean;
} & {
    newest: boolean;
} & {
    packageManager: string;
} & {
    registry: string;
} & {
    silent: boolean;
} & {
    greatest: boolean;
} & {
    upgrade: boolean;
} & {
    semverLevel: string;
} & {
    removeRange: boolean;
} & {
    dedupe: boolean;
}>;
export default setupNcuToYargs;
