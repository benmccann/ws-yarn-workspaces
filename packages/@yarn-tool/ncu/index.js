"use strict";
/**
 * Created by user on 2020/6/12.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
(0, tslib_1.__exportStar)(require("./lib/types"), exports);
(0, tslib_1.__exportStar)(require("./lib/store"), exports);
(0, tslib_1.__exportStar)(require("./lib/cli"), exports);
(0, tslib_1.__exportStar)(require("./lib/remote"), exports);
(0, tslib_1.__exportStar)(require("./lib/util"), exports);
(0, tslib_1.__exportStar)(require("./lib/options"), exports);
(0, tslib_1.__exportStar)(require("./lib/update"), exports);
exports.default = exports;
//# sourceMappingURL=index.js.map