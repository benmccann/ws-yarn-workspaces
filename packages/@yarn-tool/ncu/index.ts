/**
 * Created by user on 2020/6/12.
 */

export * from './lib/types';
export * from './lib/store';
export * from './lib/cli';
export * from './lib/remote';
export * from './lib/util';
export * from './lib/options';
export * from './lib/update';

export default exports as typeof import('./index');
