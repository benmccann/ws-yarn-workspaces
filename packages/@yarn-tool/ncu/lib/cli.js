"use strict";
/**
 * Created by user on 2020/6/12.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupNcuToYargs = void 0;
/**
 * @see https://github.com/raineorshine/npm-check-updates/blob/main/src/cli-options.ts
 */
function setupNcuToYargs(yargs) {
    return yargs
        .option('dep', {
        desc: `check only a specific section(s) of dependencies: prod|dev|peer|optional|bundle (comma-delimited)`,
        string: true,
    })
        .option('minimal', {
        alias: ['m'],
        desc: `do not upgrade newer versions that are already satisfied by the version range according to semver`,
        boolean: true,
    })
        .option('newest', {
        alias: ['n'],
        desc: `find the newest versions available instead of the latest stable versions`,
        boolean: true,
    })
        .option('packageManager', {
        alias: ['p'],
        desc: `npm (default) or bower`,
        default: 'npm',
        string: true,
    })
        .option('registry', {
        alias: ['r'],
        desc: `specify third-party npm registry`,
        string: true,
    })
        .option('silent', {
        alias: ['s'],
        desc: `don't output anything (--loglevel silent)`,
        boolean: true,
    })
        .option('greatest', {
        alias: ['g'],
        desc: `find the highest versions available instead of the latest stable versions`,
        boolean: true,
    })
        .option('upgrade', {
        alias: ['u'],
        desc: `overwrite package file`,
        boolean: true,
    })
        .option('semverLevel', {
        desc: `find the highest version within "major" or "minor"`,
        string: true,
    })
        .option('removeRange', {
        desc: `remove version ranges from the final package version`,
        boolean: true,
    })
        .option('dedupe', {
        desc: `remove upgrade module from resolutions`,
        boolean: true,
        default: true,
    })
        .option('filter', {
        desc: `Include only package names matching the given string, wildcard, glob, comma-or-space-delimited list, /regex/, or predicate function.`,
        array: true,
    });
}
exports.setupNcuToYargs = setupNcuToYargs;
exports.default = setupNcuToYargs;
//# sourceMappingURL=cli.js.map