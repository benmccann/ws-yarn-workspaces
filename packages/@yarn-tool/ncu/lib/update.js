"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.npmCheckUpdates = exports.checkResolutionsUpdate = void 0;
/**
 * Created by user on 2020/6/12.
 */
const npm_check_updates_1 = require("npm-check-updates");
const yarnlock_1 = require("@yarn-tool/yarnlock");
const bluebird_1 = __importDefault(require("bluebird"));
const util_1 = require("./util");
const semver_1 = __importDefault(require("semver"));
const remote_1 = require("./remote");
const options_1 = require("./options");
const table_1 = require("@yarn-tool/table");
const npm_package_arg_1 = __importDefault(require("npm-package-arg"));
const queryVersion_1 = require("@yarn-tool/pkg-version-query/lib/queryVersion");
const pkg_version_query_1 = require("@yarn-tool/pkg-version-query");
const parseSimpleSemVer_1 = require("@lazy-node/semver-simple-parse/lib/parseSimpleSemVer");
const mergeSimpleSemVer_1 = require("@lazy-node/semver-simple-parse/lib/mergeSimpleSemVer");
const stringifySimpleSemVer_1 = require("@lazy-node/semver-simple-parse/lib/stringifySimpleSemVer");
function checkResolutionsUpdate(resolutions, yarnlock_old_obj, options) {
    return bluebird_1.default.resolve()
        .then(async function () {
        if (typeof yarnlock_old_obj === 'string') {
            yarnlock_old_obj = yarnlock_1.parse(yarnlock_old_obj);
        }
        let result = yarnlock_1.filterResolutions({
            resolutions,
        }, yarnlock_old_obj);
        let deps = await remote_1.queryRemoteVersions(resolutions, options);
        //console.dir(deps);
        let deps2 = util_1.keyObjectToPackageMap(deps, true);
        let deps3 = Object.values(deps)
            .reduce(function (a, b) {
            a[b.name] = b;
            return a;
        }, {});
        let yarnlock_new_obj = {
            ...yarnlock_old_obj,
        };
        let update_list = [];
        let yarnlock_changed = false;
        Object.entries(result.max)
            .forEach(function ([name, data]) {
            let _key2 = name + '@' + deps3[name].version_old;
            /**
             * 檢查 版本範圍是否符合 與 版本是否不相同
             */
            //					console.dir({
            //						data,
            //						deps: deps2[name],
            //					});
            if (data.value.version != null && deps2[name] != null && semver_1.default.lt(data.value.version, deps2[name]) && yarnlock_new_obj[_key2] && yarnlock_new_obj[_key2].version != data.value.version) {
                Object.keys(result.deps[name])
                    .forEach(version => {
                    let key = name + '@' + version;
                    delete yarnlock_new_obj[key];
                });
                yarnlock_changed = true;
                update_list.push(name);
            }
            else {
                if (result.installed[name].length > 1) {
                    Object.keys(result.deps[name])
                        .forEach(version => {
                        let key = name + '@' + version;
                        yarnlock_new_obj[key] = data.value;
                    });
                    yarnlock_changed = true;
                }
            }
        });
        return {
            yarnlock_old_obj,
            yarnlock_new_obj,
            update_list,
            yarnlock_changed,
            deps,
            deps2,
            deps3,
        };
    });
}
exports.checkResolutionsUpdate = checkResolutionsUpdate;
async function npmCheckUpdates(cache, ncuOptions) {
    //ncuOptions.silent = false;
    //ncuOptions.json = false;
    //ncuOptions.cli = true;
    //ncuOptions.args = [];
    //ncuOptions.loglevel = 'verbose';
    ncuOptions = options_1.npmCheckUpdatesOptions(ncuOptions);
    ncuOptions.cwd = cache.cwd;
    ncuOptions.json_new = JSON.parse(ncuOptions.packageData);
    ncuOptions.list_updated = await npm_check_updates_1.run(ncuOptions);
    let json_changed = false;
    const current = {};
    const list_updated = {};
    await bluebird_1.default
        .resolve([
        'dependencies',
        'devDependencies',
        'peerDependencies',
        'optionalDependencies',
    ])
        .each(async (key) => {
        var _a;
        const deps = (_a = ncuOptions.json_new[key]) !== null && _a !== void 0 ? _a : {};
        await bluebird_1.default
            .resolve(Object.keys(deps))
            .each(async (name) => {
            var _a;
            const version_new = ncuOptions.list_updated[name];
            const version_old = deps[name];
            if (version_new === null || version_new === void 0 ? void 0 : version_new.length) {
                if (version_old !== version_new && util_1.allowUpdateVersion(version_old)) {
                    list_updated[name] = version_new;
                    current[name] = version_old;
                    deps[name] = version_new;
                    json_changed = true;
                }
            }
            else if (!/[\s|&]/.test(version_old)) {
                let key = `${name}@${version_old}`;
                let check = npm_package_arg_1.default(key);
                let prefix = (_a = /^([\^~\s]+)/.exec(version_old)) === null || _a === void 0 ? void 0 : _a[1];
                if ((prefix === null || prefix === void 0 ? void 0 : prefix.length) && check.type === 'range') {
                    let version_new = await queryVersion_1.queryVersionWithCache(name, version_old)
                        .then(v => prefix + v)
                        .catch(e => null);
                    if ((version_new === null || version_new === void 0 ? void 0 : version_new.length) && version_new !== version_old) {
                        try {
                            let { target } = mergeSimpleSemVer_1.mergeSimpleSemVer(parseSimpleSemVer_1.parseSimpleSemVer(version_old), parseSimpleSemVer_1.parseSimpleSemVer(version_new));
                            let version = stringifySimpleSemVer_1.stringifySemverFull(target);
                            if ((version === null || version === void 0 ? void 0 : version.length) > 0) {
                                list_updated[name] = version;
                                current[name] = version_old;
                                deps[name] = version;
                                json_changed = true;
                            }
                        }
                        catch (err) { }
                    }
                }
            }
        });
    });
    await pkg_version_query_1.getCache().fsDump();
    ncuOptions.json_changed = json_changed;
    ncuOptions.list_updated = list_updated;
    ncuOptions.current = current;
    const table = table_1.toDependencyTable({
        from: ncuOptions.current,
        to: ncuOptions.list_updated,
    }).toString();
    table && console.log(`\n${table}\n`);
    return ncuOptions;
}
exports.npmCheckUpdates = npmCheckUpdates;
//# sourceMappingURL=update.js.map