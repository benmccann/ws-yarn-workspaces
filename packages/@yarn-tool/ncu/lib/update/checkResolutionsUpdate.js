"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkResolutionsUpdate = void 0;
const bluebird_1 = require("bluebird");
const util_1 = require("../util");
const semver_1 = require("semver");
const core_1 = require("@yarn-tool/yarnlock/lib/core");
const queryRemoteVersions_1 = require("../remote/queryRemoteVersions");
const yarnlock_parse_1 = require("@yarn-tool/yarnlock-parse");
const yarnlock_parsed_to_json_1 = require("@yarn-tool/yarnlock-parsed-to-json");
function checkResolutionsUpdate(resolutions, yarnlock_old_obj, options) {
    return (0, bluebird_1.resolve)()
        .then(async function () {
        let verType;
        if (typeof yarnlock_old_obj === 'string' || Buffer.isBuffer(yarnlock_old_obj)) {
            ({ verType, parsed: yarnlock_old_obj } = (0, yarnlock_parse_1._yarnLockParseRaw)(yarnlock_old_obj));
        }
        const y_old = (0, yarnlock_parse_1._yarnLockParseCore)({
            verType,
            parsed: yarnlock_old_obj,
        });
        const result = (0, core_1.filterResolutions)({
            resolutions,
        }, y_old.data);
        const deps = await (0, queryRemoteVersions_1.queryRemoteVersions)(resolutions, options);
        //console.dir(deps);
        const deps2 = (0, util_1.keyObjectToPackageMap)(deps, true);
        const deps3 = Object.values(deps)
            .reduce(function (a, b) {
            a[b.name] = b;
            return a;
        }, {});
        const data = {
            ...y_old.data,
        };
        const update_list = [];
        let yarnlock_changed = false;
        Object.entries(result.max)
            .forEach(function ([name, data]) {
            const _key2 = name + '@' + deps3[name].version_old;
            /**
             * 檢查 版本範圍是否符合 與 版本是否不相同
             */
            //					console.dir({
            //						data,
            //						deps: deps2[name],
            //					});
            if (data.value.version != null && deps2[name] != null && (0, semver_1.lt)(data.value.version, deps2[name]) && data[_key2] && data[_key2].version != data.value.version) {
                Object.keys(result.deps[name])
                    .forEach(version => {
                    const key = name + '@' + version;
                    delete data[key];
                });
                yarnlock_changed = true;
                update_list.push(name);
            }
            else {
                if (result.installed[name].length > 1) {
                    Object.keys(result.deps[name])
                        .forEach(version => {
                        const key = name + '@' + version;
                        data[key] = data.value;
                    });
                    yarnlock_changed = true;
                }
            }
        });
        const yarnlock_new_obj = (0, yarnlock_parsed_to_json_1.yarnLockParsedToRawJSON)({
            verType,
            meta: y_old.meta,
            data,
        });
        return {
            verType: verType,
            yarnlock_old_obj: yarnlock_old_obj,
            yarnlock_new_obj,
            update_list,
            yarnlock_changed,
            deps,
            deps2,
            deps3,
        };
    });
}
exports.checkResolutionsUpdate = checkResolutionsUpdate;
//# sourceMappingURL=checkResolutionsUpdate.js.map