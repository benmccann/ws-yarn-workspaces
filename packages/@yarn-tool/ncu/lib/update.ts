/**
 * Created by user on 2020/6/12.
 */
import { run as _npmCheckUpdates } from 'npm-check-updates';
import {
	IPackageMap,
	IOptionsNpmCheckUpdates,
	IVersionCacheMapValue,
	IDependency,
	EnumVersionValue2,
	EnumVersionValue,
} from './types';
import {
	IYarnLockfileParseObject,
	parse as parseYarnLock,
	filterResolutions,
	IWrapDedupeCache,
} from '@yarn-tool/yarnlock';
import Bluebird from 'bluebird';
import { keyObjectToPackageMap, allowUpdateVersion } from './util';
import semver from 'semver';
import { queryRemoteVersions } from './remote';
import { npmCheckUpdatesOptions } from './options';
import { IPackageJsonDependenciesField } from '@ts-type/package-dts/package-json';
import { toDependencyTable } from '@yarn-tool/table';
import { ITSRequireAtLeastOne } from 'ts-type';
import npmPackageArg from 'npm-package-arg';
import { queryVersionWithCache } from '@yarn-tool/pkg-version-query/lib/queryVersion';
import { getCache } from '@yarn-tool/pkg-version-query';
import { parseSimpleSemVer } from '@lazy-node/semver-simple-parse/lib/parseSimpleSemVer';
import { mergeSimpleSemVer } from '@lazy-node/semver-simple-parse/lib/mergeSimpleSemVer';
import { stringifySemverFull } from '@lazy-node/semver-simple-parse/lib/stringifySimpleSemVer';

export function checkResolutionsUpdate(resolutions: IPackageMap,
	yarnlock_old_obj: IYarnLockfileParseObject | string,
	options: Partial<IOptionsNpmCheckUpdates>,
)
{
	return Bluebird.resolve()
		.then(async function ()
		{
			if (typeof yarnlock_old_obj === 'string')
			{
				yarnlock_old_obj = parseYarnLock(yarnlock_old_obj);
			}

			let result = filterResolutions({
				resolutions,
			}, yarnlock_old_obj);

			let deps = await queryRemoteVersions(resolutions, options);

			//console.dir(deps);

			let deps2 = keyObjectToPackageMap(deps, true);

			let deps3 = Object.values(deps)
				.reduce(function (a, b)
				{
					a[b.name] = b;

					return a;
				}, {} as Record<string, IVersionCacheMapValue>)
			;

			let yarnlock_new_obj: IYarnLockfileParseObject = {
				...yarnlock_old_obj,
			};

			let update_list: string[] = [];
			let yarnlock_changed = false;

			Object.entries(result.max)
				.forEach(function ([name, data])
				{
					let _key2 = name + '@' + deps3[name].version_old;

					/**
					 * 檢查 版本範圍是否符合 與 版本是否不相同
					 */
//					console.dir({
//						data,
//						deps: deps2[name],
//					});
					if (data.value.version != null && deps2[name] != null && semver.lt(data.value.version, deps2[name]) && yarnlock_new_obj[_key2] && yarnlock_new_obj[_key2].version != data.value.version)
					{
						Object.keys(result.deps[name])
							.forEach(version =>
							{
								let key = name + '@' + version;

								delete yarnlock_new_obj[key]
							})
						;

						yarnlock_changed = true;

						update_list.push(name);
					}
					else
					{
						if (result.installed[name].length > 1)
						{
							Object.keys(result.deps[name])
								.forEach(version =>
								{

									let key = name + '@' + version;

									yarnlock_new_obj[key] = data.value;
								})
							;

							yarnlock_changed = true;
						}
					}

				})
			;

			return {
				yarnlock_old_obj,
				yarnlock_new_obj,
				update_list,
				yarnlock_changed,
				deps,
				deps2,
				deps3,
			}
		})
		;
}

export async function npmCheckUpdates<C extends IWrapDedupeCache>(cache: Partial<C>,
	ncuOptions: ITSRequireAtLeastOne<IOptionsNpmCheckUpdates, 'json_old' | 'packageData'>,
)
{
	//ncuOptions.silent = false;

	//ncuOptions.json = false;
	//ncuOptions.cli = true;

	//ncuOptions.args = [];

	//ncuOptions.loglevel = 'verbose';

	ncuOptions = npmCheckUpdatesOptions(ncuOptions);

	ncuOptions.cwd = cache.cwd;

	ncuOptions.json_new = JSON.parse(ncuOptions.packageData);

	ncuOptions.list_updated = await _npmCheckUpdates(ncuOptions) as Record<string, string>;

	let json_changed = false;

	const current: IDependency = {};
	const list_updated: IDependency = {};

	await Bluebird
		.resolve([
			'dependencies',
			'devDependencies',
			'peerDependencies',
			'optionalDependencies',
		] as IPackageJsonDependenciesField[])
		.each(async (key) =>
		{
			const deps = ncuOptions.json_new[key] ?? {};

			await Bluebird
				.resolve(Object.keys(deps))
				.each(async (name) =>
				{
					const version_new = ncuOptions.list_updated[name];
					const version_old = deps[name];

					if (version_new?.length)
					{
						if (version_old !== version_new && allowUpdateVersion(version_old))
						{
							list_updated[name] = version_new;
							current[name] = version_old;

							deps[name] = version_new;

							json_changed = true;
						}
					}
					else if (!/[\s|&]/.test(version_old))
					{
						let key = `${name}@${version_old}`

						let check = npmPackageArg(key)
						let prefix = /^([\^~\s]+)/.exec(version_old)?.[1];

						if (prefix?.length && check.type === 'range')
						{
							let version_new = await queryVersionWithCache(name, version_old)
								.then(v => prefix + v)
								.catch(e => null)
							;

							if (version_new?.length && version_new !== version_old)
							{
								try
								{
									let { target } = mergeSimpleSemVer(parseSimpleSemVer(version_old), parseSimpleSemVer(version_new));

									let version = stringifySemverFull(target)

									if (version?.length > 0)
									{
										list_updated[name] = version;
										current[name] = version_old;

										deps[name] = version;

										json_changed = true;
									}
								}
								catch (err)
								{}
							}
						}
					}

				})

		})
	;

	await getCache().fsDump();

	ncuOptions.json_changed = json_changed;
	ncuOptions.list_updated = list_updated;
	ncuOptions.current = current;

	const table = toDependencyTable({
		from: ncuOptions.current,
		to: ncuOptions.list_updated,
	}).toString();

	table && console.log(`\n${table}\n`);

	return ncuOptions;
}
