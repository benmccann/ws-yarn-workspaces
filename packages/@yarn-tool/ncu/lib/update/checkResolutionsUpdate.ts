import { IPackageMap, IOptionsNpmCheckUpdates, IVersionCacheMapValue } from '../types';
import { resolve } from 'bluebird';
import { keyObjectToPackageMap } from '../util';
import { lt } from 'semver';
import { filterResolutions } from '@yarn-tool/yarnlock/lib/core';
import { queryRemoteVersions } from '../remote/queryRemoteVersions';
import {
	EnumDetectYarnLock,
	IYarnLockDataRecord,
	IYarnLockDataRowV1,
	IYarnLockDataRowV2,
	IYarnLockSource,
} from '@yarn-tool/yarnlock-types';
import { _yarnLockParseCore, _yarnLockParseRaw } from '@yarn-tool/yarnlock-parse';
import { yarnLockParsedToRawJSON } from '@yarn-tool/yarnlock-parsed-to-json';
import { ITSExcludeEnumValue } from 'ts-type/lib/helper/record/enum';

export function checkResolutionsUpdate<T extends IYarnLockSource>(resolutions: IPackageMap,
	yarnlock_old_obj: IYarnLockSource | Buffer | string,
	options: Partial<IOptionsNpmCheckUpdates>,
)
{
	return resolve()
		.then(async function ()
		{
			let verType: ITSExcludeEnumValue<typeof EnumDetectYarnLock, 0>;

			if (typeof yarnlock_old_obj === 'string' || Buffer.isBuffer(yarnlock_old_obj))
			{
				({ verType, parsed: yarnlock_old_obj } = _yarnLockParseRaw(yarnlock_old_obj));
			}

			const y_old = _yarnLockParseCore({
				verType,
				parsed: yarnlock_old_obj,
			});

			const result = filterResolutions({
				resolutions,
			}, y_old.data);

			const deps = await queryRemoteVersions(resolutions, options);

			//console.dir(deps);

			const deps2 = keyObjectToPackageMap(deps, true);

			const deps3 = Object.values(deps)
				.reduce(function (a, b)
				{
					a[b.name] = b;

					return a;
				}, {} as Record<string, IVersionCacheMapValue>)
			;

			const data: IYarnLockDataRecord = {
				...y_old.data,
			};

			const update_list: string[] = [];
			let yarnlock_changed = false;

			Object.entries(result.max)
				.forEach(function ([name, data])
				{
					const _key2 = name + '@' + deps3[name].version_old;

					/**
					 * 檢查 版本範圍是否符合 與 版本是否不相同
					 */
//					console.dir({
//						data,
//						deps: deps2[name],
//					});
					if (data.value.version != null && deps2[name] != null && lt(data.value.version, deps2[name]) && data[_key2] && data[_key2].version != data.value.version)
					{
						Object.keys(result.deps[name])
							.forEach(version =>
							{
								const key = name + '@' + version;

								delete data[key]
							})
						;

						yarnlock_changed = true;

						update_list.push(name);
					}
					else
					{
						if (result.installed[name].length > 1)
						{
							Object.keys(result.deps[name])
								.forEach(version =>
								{

									const key = name + '@' + version;

									data[key] = data.value;
								})
							;

							yarnlock_changed = true;
						}
					}

				})
			;

			const yarnlock_new_obj = yarnLockParsedToRawJSON({
				verType,
				meta: y_old.meta,
				data,
			}) as T

			return {
				verType: verType as ITSExcludeEnumValue<typeof EnumDetectYarnLock, 0>,
				yarnlock_old_obj: yarnlock_old_obj as T,
				yarnlock_new_obj,
				update_list,
				yarnlock_changed,
				deps,
				deps2,
				deps3,
			}
		})
		;
}
