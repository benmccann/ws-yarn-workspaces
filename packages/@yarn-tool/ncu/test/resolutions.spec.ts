//@noUnusedParameters:false

import { join } from 'path';
import { __TEST_YARNLOCK } from '../../../../__root_ws';
import { readFileSync } from 'fs';
import { checkResolutionsUpdate } from '../lib/update/checkResolutionsUpdate';
import { pathExistsSync } from 'fs-extra';
import { _forEachVersionTags } from '../../../../test/lib/forEachVersionTags';
import { EnumDetectYarnLock } from '@yarn-tool/yarnlock-types';

beforeAll(async () =>
{

});

describe(`checkResolutionsUpdate`, () =>
{
	const dir = join(__TEST_YARNLOCK, 'ncu');

	_forEachVersionTags({
		only: [
			EnumDetectYarnLock.v1,
		],
	}).forEach(ver =>
	{
		const file = join(dir, ver, 'yarn.lock');

		pathExistsSync(file) && test(ver, async () =>
		{
			const name = "@types/node" as const;

			const content = readFileSync(file).toString();

			let actual = await checkResolutionsUpdate({
				[name]: "^1.0.10",
			}, content, {});

			const row = {
				name,
				version_new: expect.any(String),
			};

			console.dir(actual.deps3);
			console.dir(actual.yarnlock_changed);
			console.dir(actual.update_list);

			expect(actual).toMatchSnapshot({
				deps: [row],
				deps2: {
					[name]: row.version_new,
				},
				deps3: {
					[name]: row,
				},
				update_list: expect.any(Array),
				yarnlock_changed: expect.any(Boolean),
				yarnlock_new_obj: expect.anything(),
				yarnlock_old_obj: expect.anything(),
			});

		});
	});

})
