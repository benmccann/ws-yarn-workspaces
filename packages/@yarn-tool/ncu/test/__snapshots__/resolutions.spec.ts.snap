// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkResolutionsUpdate v1 1`] = `
Object {
  "deps": Array [
    Object {
      "name": "@types/node",
      "versionTarget": "latest",
      "version_new": Any<String>,
      "version_old": "^1.0.10",
    },
  ],
  "deps2": Object {
    "@types/node": Any<String>,
  },
  "deps3": Object {
    "@types/node": Object {
      "name": "@types/node",
      "versionTarget": "latest",
      "version_new": Any<String>,
      "version_old": "^1.0.10",
    },
  },
  "update_list": Any<Array>,
  "verType": 1,
  "yarnlock_changed": Any<Boolean>,
  "yarnlock_new_obj": Anything,
  "yarnlock_old_obj": Anything,
}
`;

exports[`checkResolutionsUpdate v3 1`] = `
Object {
  "deps": Array [
    Object {
      "name": "@types/node",
      "versionTarget": "latest",
      "version_new": Any<String>,
      "version_old": "^1.0.10",
    },
  ],
  "deps2": Object {
    "@types/node": Any<String>,
  },
  "deps3": Object {
    "@types/node": Object {
      "name": "@types/node",
      "versionTarget": "latest",
      "version_new": Any<String>,
      "version_old": "^1.0.10",
    },
  },
  "update_list": Any<Array>,
  "verType": 3,
  "yarnlock_changed": Any<Boolean>,
  "yarnlock_new_obj": Anything,
  "yarnlock_old_obj": Anything,
}
`;
